### ISSUE TITLE
Possible `Lint/UselessAssignment` false negative

### ISSUE THREAD

**Issue by vlad-pisanov** [rubocop/rubocop#11786]
Consider:

```ruby
unused = 1
arr.each { unused = 2 }
```

Either line in isolation triggers `Lint/UselessAssignment`, but when taken together, no offense is registered.

In contrast, this triggers `Lint/UselessAssignment` correctly:
```ruby
unused = 1   # 👮‍♂️ Useless assignment to variable - unused 
if cond
  unused = 2 # 👮‍♂️ Useless assignment to variable - unused
end
```

### Rubocop version
```
1.48.1 (using Parser 3.2.1.1, rubocop-ast 1.27.0, running on ruby 2.6.9) [x86_64-darwin21]
  - rubocop-minitest 0.29.0
  - rubocop-performance 1.16.0
  - rubocop-rails 2.18.0
```

**Comment by bbatsov** [comment]
Reopened — still reproduces on master (1.88): `unused = 1; arr.each { unused = 2 }` registers no offense, even though `unused = 1` on its own is flagged. The dead store is missed once the variable is reassigned inside a block.
